internal/controller/awsaccount_controller.go

```diff
@@ -0,0 +1,109 @@
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/sethvargo/go-password/password"
+
+	kuadrav1 "github.com/Kuadrant/kuadra/api/v1"
+	"github.com/Kuadrant/kuadra/pkg/aws"
+)
+
+// AwsAccountReconciler reconciles a AwsAccount object
+type AwsAccountReconciler struct {
+	client.Client
+	Scheme     *runtime.Scheme
+	IamWrapper aws.IamWrapper
+}
+
+//+kubebuilder:rbac:groups=kuadra.kuadrant.io,resources=awsaccounts,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=kuadra.kuadrant.io,resources=awsaccounts/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=kuadra.kuadrant.io,resources=awsaccounts/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the AwsAccount object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.4/pkg/reconcile
+func (r *AwsAccountReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	// Get AwsAccount object object
+	var awsAccount kuadrav1.AwsAccount
+	if err := r.Get(ctx, req.NamespacedName, &awsAccount); err != nil {
+		log.Error(err, "unable to fetch AwsAccount")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Create user if status.UserCreated is false
+	if !awsAccount.Status.UserCreated {
+		user, err := r.IamWrapper.CreateUser(awsAccount.Spec.UserName)
+		if err != nil {
+			log.Error(err, "unable to create IAM user")
+			return ctrl.Result{}, err
+		}
+
+		pass, err := password.Generate(20, 3, 3, false, true)
+		if err != nil {
+			log.Error(err, "unable to generate password")
+			return ctrl.Result{}, err
+		}
+
+		_, err = r.IamWrapper.CreateLoginProfile(pass, awsAccount.Spec.UserName, true)
+		if err != nil {
+			log.Error(err, "unable to create login profile")
+			return ctrl.Result{}, err
+		}
+
+		accessKey, err := r.IamWrapper.CreateAccessKeyPair(awsAccount.Spec.UserName)
+		if err != nil {
+			log.Error(err, "unable to create access key")
+			return ctrl.Result{}, err
+		}
+
+		// TODO: Save these to a secret in user's namespace
+		log.V(1).Info("Credentials", "password", pass, "access key ID:", accessKey.AccessKeyId, "Access key secret", accessKey.SecretAccessKey)
+
+		// TODO: Maybe we should use more detailed statuses here, such as "not created", "missing login profile", "missing access key", "created"
+		awsAccount.Status.UserCreated = true
+		if err = r.Status().Update(ctx, &awsAccount); err != nil {
+			log.Error(err, "unable to update awsAccount status")
+			return ctrl.Result{}, err
+		}
+		log.V(1).Info("Created user", "User:", user)
+	}
+
+	return ctrl.Result{}, nil
```

Comment:
```suggestion
	// Create user if status.UserCreated is false
	if awsAccount.Status.UserCreated {
		return ctrl.Result{}, nil	
	}
	
	user, err := r.IamWrapper.CreateUser(awsAccount.Spec.UserName)
	if err != nil {
		log.Error(err, "unable to create IAM user")
		return ctrl.Result{}, err
	}

	pass, err := password.Generate(20, 3, 3, false, true)
	if err != nil {
		log.Error(err, "unable to generate password")
		return ctrl.Result{}, err
	}

	_, err = r.IamWrapper.CreateLoginProfile(pass, awsAccount.Spec.UserName, true)
	if err != nil {
		log.Error(err, "unable to create login profile")
		return ctrl.Result{}, err
	}

	accessKey, err := r.IamWrapper.CreateAccessKeyPair(awsAccount.Spec.UserName)
	if err != nil {
		log.Error(err, "unable to create access key")
		return ctrl.Result{}, err
	}

	// TODO: Save these to a secret in user's namespace
	log.V(1).Info("Credentials", "password", pass, "access key ID:", accessKey.AccessKeyId, "Access key secret", accessKey.SecretAccessKey)

	// TODO: Maybe we should use more detailed statuses here, such as "not created", "missing login profile", "missing access key", "created"
	awsAccount.Status.UserCreated = true
	if err = r.Status().Update(ctx, &awsAccount); err != nil {
		log.Error(err, "unable to update awsAccount status")
		return ctrl.Result{}, err
	}
	log.V(1).Info("Created user", "User:", user)
	return ctrl.Result{}, nil
```

Just a thing that I find good practice when writing code is what I think it's called _early return_. Instead of wrapping that big block of code into an `if`, invert the condition of the `if` and `return` (or `continue` if you're within a loop). That makes the code a bit more readable by avoiding indentation, which in more complex cases can result in lots of nested blocks

Reply:
Thanks @sergioifg94, thinking about this, I was wondering would it make sense for us to maybe split this block up into smaller `if` blocks and have more regular and detailed status updates?

We're currently just using the `Status.UserCreated` boolean for the entire block, but I was thinking we could replace this with something like `Status.Account` which would hold more detailed status such as "creating user", "creating login profle", "creating access key", etc. This status could then be used to determine which blocks will run and might make the reconciler a bit more robust if any of the requests to AWS were to fail.

@Ygnas